pkg/envoyfilter/controller.go

```diff
@@ -0,0 +1,265 @@
+	// debounceAfter is the delay added to events to wait after a registry event for debouncing.
+	// This will delay the push by at least this interval, plus the time getting subsequent events.
+	// If no change is detected the push will happen, otherwise we'll keep delaying until things settle.
+	debounceAfter = 100 * time.Millisecond
+
+	// debounceMax is the maximum time to wait for events while debouncing.
+	// Defaults to 10 seconds. If events keep showing up with no break for this time, we'll trigger a push.
+	debounceMax = 10 * time.Second
+
+	// configRootNS is the root config root namespace
+	configRootNS = "istio-system"
+
+	// aerakiFieldManager is the FileldManager for Aeraki CRDs
+	aerakiFieldManager = "Aeraki"
+)
+
+var (
+	controllerLog = log.RegisterScope("envoyfilter-controller", "envoyfilter-controller debugging", 0)
+)
+
+// Controller contains the runtime configuration for the envoyFilter controller.
+type Controller struct {
+	configStore istiomodel.ConfigStore
+	generators  map[protocol.Instance]Generator
+	// Sending on this channel results in a push.
+	pushChannel chan istiomodel.Event
+}
+
+// NewController creates a new controller instance based on the provided arguments.
+func NewController(store istiomodel.ConfigStore, generators map[protocol.Instance]Generator) *Controller {
+	controller := &Controller{
+		configStore: store,
+		generators:  generators,
+		pushChannel: make(chan istiomodel.Event, 100),
+	}
+	return controller
+}
+
+// Run until a signal is received, this function won't block
+func (s *Controller) Run(stop <-chan struct{}) {
+	go func() {
+		s.mainLoop(stop)
+	}()
+}
+
+func (s *Controller) mainLoop(stop <-chan struct{}) {
+	var timeChan <-chan time.Time
+	var startDebounce time.Time
+	var lastResourceUpdateTime time.Time
+	pushCounter := 0
+	debouncedEvents := 0
+
+	for {
+		select {
+		case <-stop:
+			break
+		case e := <-s.pushChannel:
+			controllerLog.Debugf("Receive event from push chanel : %v", e)
+			lastResourceUpdateTime = time.Now()
+			if debouncedEvents == 0 {
+				controllerLog.Debugf("This is the first debounced event")
+				startDebounce = lastResourceUpdateTime
+			}
+			timeChan = time.After(debounceAfter)
+			debouncedEvents++
+		case <-timeChan:
+			controllerLog.Debugf("Receive event from time chanel")
+			eventDelay := time.Since(startDebounce)
+			quietTime := time.Since(lastResourceUpdateTime)
+			// it has been too long since the first debounced event or quiet enough since the last debounced event
+			if eventDelay >= debounceMax || quietTime >= debounceAfter {
+				if debouncedEvents > 0 {
+					pushCounter++
+					controllerLog.Infof("Push debounce stable[%d] %d: %v since last change, %v since last push",
+						pushCounter, debouncedEvents, quietTime, eventDelay)
+					s.pushEnvoyFilters2APIServer()
+					debouncedEvents = 0
+				}
+			} else {
+				timeChan = time.After(debounceAfter - quietTime)
+			}
+		}
+	}
+}
+
+func (s *Controller) pushEnvoyFilters2APIServer() {
```

Comment:
Maybe should introduce a `Reconcile` like mechanism to ensure the generated EnvoyFilters will be managed as expect.

Reply:
It's a bit tricky, we'll fix this in another PR.